rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+
+If this is perceived to be a problem, we could decide to do away with some of
+the complexity by initially focusing on a restricted subset of this proposal
+which only supports `Relaxed` loads and stores, and saving more complex atomic
+operations as a future extension.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This effort seems worthwhile because it simultaneously eliminates two well-known
+and unnecessary volatile footguns (data races and tearing) and opens interesting
+new possibilities in interprocess communication.
+
+But although this feature may seem simple, its design space is actually
+remarquably large, and a great many alternatives were considered before reaching
+the current design proposal. Here are some design knobs that were explored:
+
+## Extending `AtomicXyz`
+[extending-atomicxyz]: #extending-atomicxyz
+
+Atomic volatile operations could be made part of the Atomic wrapper types.
+However, they cannot be just another atomic memory ordering, due to the fact
+that atomic ops take an `&self` (which is against the spirit of volatile as it
+is tagged with `dereferenceable` at the LLVM layer).
+
+Instead, one would need to add more methods to atomic wrapper types, which take
+a pointer as a self-parameter. This API inconsistency was felt to be
+unnecessarily jarring to users, not to mention that potentially having a copy of
+each atomic memory operation with a `_volatile` prefix would reduce the
+readability of the `AtomicXyz` types' API documentation.
+
+In addition, it is extremely common to want either all operations on a memory
+location to be volatile, or none of them. Providing separate wrapper types
+helps enforce this very common usage pattern at the API level.
+
+## Self-type or not self-type
+[self-type]: #self-type
+
+Instead of using `arbitrary_self_types` to get `a.load(Ordering::Relaxed)`
+method syntax on pointer-like objects, one could instead provide the volatile
+operations as inherent methods, i.e. `VolatileU8::load(a, Ordering::Relaxed)`.
+
+This has the advantage of avoiding coupling this feature to another unstable
+feature. But it has the drawback of being incredibly verbose. Typing the same
+volatile type name over and over again in a complex shared memory transaction
+would certainly get old and turn annoying quickly, and we don't want anger to
+distract low-level developers from the delicate task of implementing the kind
+of subtle algorithm that requires volatile operations.
+
+## `NonNull<T>` vs `*mut T` vs `*const T` vs other
+[pointers]: #pointers
+
+It is pretty clear that volatile operations cannot be expressed through `&self`
+Rust references, as these provide the compiler with a licence to inject
+arbitrary loads from the memory region (or even stores, for register spills, if
+the region is determined to be unobservable to other threads). This would be
+incompatible with the stated goal of precisely controlling memory accesses.
```

Comment:
This is the status quo, but there is a lot of interest in changing that. In particular, the fact that we add `dereferencable` is [a problem even for non-volatile atomics](https://github.com/rust-lang/rust/issues/55005), so both should likely use the same approach to handle that problem. If that approach is `NonNull<Self>`, then the `AtomicX` APIs should also provide that.

Reply:
I feel the linked issue is a bit different, in the sense that the root cause there AIUI is that LLVM's current `dereferenceable` semantics are incompatible with freeing of the dereferenceable memory block. If this analysis is correct, this has further-reaching implication than atomics alone, it probably means that Rust references to memory that can be freed (e.g. the insides of a Box) are never okay, and thus that a weaker form of `dereferenceable` which is compatible with freeing sounds like the proper fix.

The MMIO issue seems different to me, in the sense that the problem is not that `dereferenceable` semantics are not okay in some case, but more strongly that it is never, ever okay for the compiler to insert a load to an MMIO region without being told to by the programmer, as loads can have arbitrary side-effects.

That being said, even if I don't agree with the linked issue being similar to the present one (besides both originating from `dereferenceable` semantics), I agree with your statement there that `&UnsafeCell<T>` being `dereferenceable` does not feel like a great idea overall. For another example, given that the core purpose of this annotation is to allow prefetch loads, it is not 100% clear in my mind why it doesn't allow the compiler's optimizer to introduce read-write data races which weren't present in the original code... even though it should hopefully have no good reason to if the memory isn't read in a properly synchronized way later on.

Anyway, all that long comment to say that if you feel that the `dereferenceable` discussion is moving forward, I think this RFC should probably be blocked on it because e.g. removing `dereferenceable` on `&UnsafeCell` would have a major positive impact on its ergonomics (yay, no more raw pointers!). But for that, I need a place where the discussion is ongoing to point to, and this issue does not sound like the greatest. Any better suggestion?

Reply:
> I feel the linked issue is a bit different, in the sense that the root cause there AIUI is that LLVM's current dereferenceable semantics are incompatible with freeing of the dereferenceable memory block. If this analysis is correct, this has further-reaching implication than atomics alone, it probably means that Rust references to memory that can be freed (e.g. the insides of a Box) are never okay,

I didn't realize (or forgot) that we also put `dereferencable` on boxes... that makes no sense at all.

>  and thus that a weaker form of dereferenceable which is compatible with freeing sounds like the proper fix.

Well, that is one of many options. Or we could pick a fix that also works for `volatile`. I am not sure which is better, so it feels premature to duplicate the entire `Atomic*` API for that.

OTOH, this only blocks stabilization, not initial experimentation.

> it is not 100% clear in my mind why it doesn't allow the compiler's optimizer to introduce read-write data races which weren't present in the original code... even though it should hopefully have no good reason to if the memory isn't read in a properly synchronized way later on.

Note that this optimization happens *in LLVM*, so it uses LLVM semantics. Under LLVM semantics, read-write races are okay and the read returns `undef`. Thus, as long as the code only *looks* at the prefetch loads in cases where the original code did an actual read (and hence there is no data race), there is no problem. This is in fact a key motivation for LLVM's choice of data race semantics!

> Anyway, all that long comment to say that if you feel that the dereferenceable discussion is moving forward

Well, nothing much ha moved in that issue for a bit. However, there is a lot of recent discussion in the embedded space around this issue, see https://github.com/rust-embedded/wg/pull/387.

Reply:
I agree that ultimately, this is more of a blocker for stabilization than unstable experimentation.

So I think the next step here is for me to carefully read through all those `dereferenceable` discussions again and try to extract some sort of small summary that can be added to the "unresolved questions" section of this RFC. Right now, the main threads I'm aware of are...

- The Arc unsoundness thread ( https://github.com/rust-lang/rust/issues/55005 ) and recursive Rust issues that it points to.
- The rust-embedded thread https://github.com/rust-embedded/wg/pull/387
- The Zulip thread that was freshly opened by @pnkfelix and @rkruppe on wg-llvm

If I'm forgetting another big one, feel free to ping me about it.

Reply:
Another good entry point into the LLVM side of the discussion seems to be https://github.com/rust-lang/rust/issues/66600#issuecomment-557531234

Reply:
Ok, in the end I just don't manage to collect the time for a full write-up here so I just left links around so that people with more time on their hand can do so more easily ;)